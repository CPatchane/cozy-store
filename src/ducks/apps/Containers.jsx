import { connect } from 'react-redux'

import {
  fetchApps,
  fetchInstalledApps,
  fetchLatestApp,
  uninstallApp,
  getInstalledApps,
  getRegistryApps,
  installAppFromRegistry,
  installApp
} from './index'

import DiscoverComponent from './components/Discover'
import HiddenInstallerViewComponent from './components/HiddenInstallerView'
import { translate } from 'cozy-ui/react/I18n'
import MyApplicationsComponent from './components/MyApplications'

const mapStateToProps = (state, ownProps) => ({
  apps: getRegistryApps(state),
  installedApps: getInstalledApps(state),
  isFetching: state.apps.isFetching,
  isAppFetching: state.apps.isAppFetching,
  isInstalling: state.apps.isInstalling,
  actionError: state.apps.actionError,
  fetchError: state.apps.fetchError
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchApps: () => dispatch(fetchApps(ownProps.lang)),
  fetchInstalledApps: () => dispatch(fetchInstalledApps(ownProps.lang)),
  installApp: (appSlug, appType, channel) =>
    dispatch(installAppFromRegistry(appSlug, appType, channel)),
  uninstallApp: (appSlug, appType) => dispatch(uninstallApp(appSlug, appType)),
  updateApp: (appSlug, appType, channel) =>
    dispatch(installAppFromRegistry(appSlug, appType, channel, true)),
  fetchLatestApp: (slug, channel) => dispatch(fetchLatestApp(slug, channel)),
  // for the hidden installer only
  installUsingInstaller: (appSlug, appType, source, isUpdate) =>
    dispatch(installApp(appSlug, appType, source, isUpdate)).catch(() => {
      dispatch({
        type: 'SEND_LOG_FAILURE',
        alert: {
          message: `HiddenInstallerView.${
            isUpdate ? 'update' : 'install'
          }_error`,
          messageData: { slug: appSlug },
          level: 'error'
        }
      })
    })
})

// translate last to pass the lang property to fetchApps()
export const Discover = translate()(connect(mapStateToProps, mapDispatchToProps)(
  DiscoverComponent
))

export const HiddenInstallerView = translate()(connect(mapStateToProps, mapDispatchToProps)(
  HiddenInstallerViewComponent
))

export const MyApplications = translate()(connect(mapStateToProps, mapDispatchToProps)(
  MyApplicationsComponent
))
