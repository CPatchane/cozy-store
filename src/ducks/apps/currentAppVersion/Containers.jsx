import { connect } from 'react-redux'

import { fetchLastAppVersion } from './index'

import InstallModalComponent from '../components/InstallModal'
import ApplicationPageComponent from '../components/ApplicationPage'

const mapStateToProps = (state, ownProps) => ({
  currentAppVersion: state.currentAppVersion.version,
  isVersionFetching: state.currentAppVersion.isFetching,
  versionError: state.currentAppVersion.error
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchLastAppVersion: (appSlug, channel) => dispatch(fetchLastAppVersion(appSlug, channel))
})

export const InstallModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(InstallModalComponent)

export const ApplicationPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(ApplicationPageComponent)
