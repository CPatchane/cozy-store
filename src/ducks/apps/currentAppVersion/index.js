/* eslint-env browser */
/* global cozy */

import { combineReducers } from 'redux'

import {
  UnavailableRegistryException,
  NotFoundException
} from '../../../lib/exceptions'

const FETCH_APP_VERSION = 'FETCH_APP_VERSION'
const FETCH_APP_VERSION_SUCCESS = 'FETCH_APP_VERSION_SUCCESS'
const FETCH_APP_VERSION_FAILURE = 'FETCH_APP_VERSION_FAILURE'

export const version = (state = null, action) => {
  switch (action.type) {
    case FETCH_APP_VERSION_SUCCESS:
      return action.version
    default:
      return state
  }
}

export const isFetching = (state = false, action) => {
  switch (action.type) {
    case FETCH_APP_VERSION:
      return true
    case FETCH_APP_VERSION_SUCCESS:
    case FETCH_APP_VERSION_FAILURE:
      return false
    default:
      return state
  }
}

export const error = (state = null, action) => {
  switch (action.type) {
    case FETCH_APP_VERSION_FAILURE:
      return action.error
    default:
      return state
  }
}

export const currentAppVersionReducers = combineReducers({
  version,
  error,
  isFetching
})

export function fetchLastAppVersion (appSlug, channel = 'stable') {
  return (dispatch, getState) => {
    dispatch({type: FETCH_APP_VERSION})
    return cozy.client.fetchJSON('GET', `/registry/${appSlug}/${channel}/latest`)
    .then(version => {
      return dispatch({type: FETCH_APP_VERSION_SUCCESS, version})
    })
    .catch(error => {
      let registryError = new UnavailableRegistryException()
      if (error.status === 404) registryError = new NotFoundException()
      dispatch({type: FETCH_APP_VERSION_FAILURE, error: registryError})
      console.error(error)
    })
  }
}
