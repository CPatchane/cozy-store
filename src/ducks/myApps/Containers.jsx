import { connect } from 'react-redux'

import { fetchApps, fetchMyApps, installApp, uninstallApp, getInstalledApps } from './index'

import MyApplicationsComponent from './components/MyApplications'
import DiscoverComponent from './components/Discover'

const mapStateToProps = (state, ownProps) => ({
  apps: state.apps.list,
  myApps: getInstalledApps(state),
  isFetching: state.apps.isFetching,
  error: state.apps.error
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchApps: () => dispatch(fetchApps()),
  fetchMyApps: () => dispatch(fetchMyApps()),
  installApp: (appSlug, channel) => dispatch(installApp(appSlug, channel)),
  uninstallApp: (appSlug) => dispatch(uninstallApp(appSlug))
})

export const MyApplications = connect(
  mapStateToProps,
  mapDispatchToProps
)(MyApplicationsComponent)

export const Discover = connect(
  mapStateToProps,
  mapDispatchToProps
)(DiscoverComponent)
