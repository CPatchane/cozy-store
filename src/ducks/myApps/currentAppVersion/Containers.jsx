import { connect } from 'react-redux'

import { fetchLastAppVersion } from './index'

import InstallModalComponent from '../components/InstallModal'

const mapStateToProps = (state, ownProps) => ({
  currentAppVersion: state.currentAppVersion.version,
  isVersionFetching: state.currentAppVersion.isFetching,
  versionError: state.currentAppVersion.error
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchLastAppVersion: (appSlug, channel) => dispatch(fetchLastAppVersion(appSlug, channel))
})

export const InstallModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(InstallModalComponent)
