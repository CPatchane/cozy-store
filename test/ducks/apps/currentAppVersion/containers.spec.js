'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'

import { InstallModal } from 'ducks/apps/currentAppVersion/Containers'

const mockStore = configureStore()

describe('currentAppVersion Containers (connected components):', () => {
  const initialState = {
    currentAppVersion: {
      version: null,
      isFetching: false,
      error: null
    }
  }

  it('InstallModal should be rendered correctly at initial store', () => {
    const store = mockStore(initialState)
    const component = shallow(
      <InstallModal store={store} />
    ).node
    expect(component).toMatchSnapshot()
  })

  it('InstallModal should dispatch one action on fetchLastAppVersion', () => {
    const store = mockStore(initialState)
    store.dispatch = jest.fn(() => Promise.resolve())
    const component = shallow(
      <InstallModal store={store} />
    )
    component.props().fetchLastAppVersion()
    expect(store.dispatch.mock.calls.length).toBe(1)
    expect(store.dispatch.mock.calls[0][0]).toBeInstanceOf(Function)
  })
})
