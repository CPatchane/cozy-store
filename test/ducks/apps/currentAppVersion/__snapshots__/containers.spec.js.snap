// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currentAppVersion Containers (connected components): InstallModal should be rendered correctly at initial store 1`] = `
<_translate
  currentAppVersion={null}
  fetchLastAppVersion={[Function]}
  isVersionFetching={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
  versionError={null}
/>
`;
